buildSrc/src/main/kotlin/androidx/build/ftl/GCloudCLIWrapper.kt

```diff
@@ -44,53 +46,103 @@ internal class GCloudCLIWrapper(
-        val result = execOperations.exec {
-            it.commandLine("which", "gcloud")
+        val errorOutput = ByteArrayOutputStream()
+        val execResult = execOperations.exec {
+            it.executable = gcloud
+            it.args = params.toList() + "--format=json"
             it.standardOutput = output
+            it.errorOutput = errorOutput
             it.isIgnoreExitValue = true
         }
-        if (result.exitValue != 0) {
-            throw GradleException(
-                """
-                Unable to find gcloud CLI executable.
-                `which gcloud` returned exit code ${result.exitValue}.
-                Make sure gcloud CLI is installed, authenticated and is part of your PATH.
-                See https://cloud.google.com/sdk/gcloud for installation instructions.
-                """.trimIndent()
-            )
+        if (execResult.exitValue != 0) {
+            System.err.println("GCloud command failed: ${errorOutput.toString(Charsets.UTF_8)}")
         }
-        output.toString(Charsets.UTF_8).trim()
+        // still try to parse the because when it fails (e.g. test failure), it returns a non-0
+        // exit code but still prints the output. We are interested in the output.
+        val commandOutput = output.toString(Charsets.UTF_8)
+        return gson.parse(commandOutput)
     }
 
-    private inline fun <reified T> executeGcloud(
+    private fun execGsUtil(
         vararg params: String
-    ): T {
+    ): String {
         val output = ByteArrayOutputStream()
         execOperations.exec {
-            it.executable = executable
-            it.args = params.toList() + "--format=json"
+            it.executable = gsutil
+            it.args = params.toList()
             it.standardOutput = output
         }
-        val commandOutput = output.toString(Charsets.UTF_8)
-        return gson.parse(commandOutput)
+        return output.toString(Charsets.UTF_8)
     }
 
     /**
      * https://cloud.google.com/sdk/gcloud/reference/firebase/test/android/run
      */
     fun runTest(
-        testedApk: File,
-        testApk: File
+        params: RunTestParameters
     ): List<TestResult> {
-        return executeGcloud(
+        val testResults = executeGcloud<List<TestResult>>(
             "firebase", "test", "android", "run",
             "--type", "instrumentation",
-            "--test", testApk.canonicalPath,
-            "--app", testedApk.canonicalPath,
-            "--num-flaky-test-attempts", "3",
+            "--test", params.testApk.canonicalPath,
+            "--app", params.testedApk.canonicalPath,
+            "--num-flaky-test-attempts", "2",
+            "--results-bucket=${params.bucketName}",
+            "--results-dir=${params.resultsBucketDir}",
+            "--results-history-name=${params.projectPath}"
+        )
+        // copy the test results from the bucket to the build folder
+        val localFolder = params.resultsLocalDir
```

Comment:
Why create the extra variable? Can we use this directly in execGsUtil?

Reply:
yea we can. :) fixed.